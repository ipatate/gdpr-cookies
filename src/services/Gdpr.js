// @flow
import merge from 'deepmerge';
import GdprCookie from './GdprCookie';
import GdprObservable from './GdprObservable';

export default class Gdpr {
  options: OptionsGdpr = {
    name: 'gdpr_cookie',
    keepCookies: ['plop'],
    types: ['ads', 'stats'],
  };
  GdprObservable: GdprObservable;
  cookie: GdprCookie;
  activated: Object;

  /**
   * @description constructor !
   * @param {object} options
   * @return {void}
   */
  constructor(options: ?OptionsGdpr = {name: 'gdpr_cookie'}): void {
    this.options = merge.all([this.options, options]);
    this.cookie = new GdprCookie(this.options.name);
    // init activated object and save in cookie
    this.initCookie();
    this.GdprObservable = new GdprObservable(
      this.getGlobalGdpr(),
      this.options.types,
    );
  }

  /**
   * @description get global _gdpr array
   * @return {array}
   */
  getGlobalGdpr(): ObserverGdpr {
    const _gdpr = global._gdpr || [];
    return _gdpr;
  }

  /**
   * @description get value in options by name
   * @param {string} name
   * @return {void}
   */
  getOption(name: string): void {
    if (this.options[name] !== undefined) {
      return this.options[name];
    }
    return undefined;
  }

  /**
   * @description each activated for active or unactive service
   * @return {void}
   */
  activeService(): void {
    for (const key in this.activated) {
      const value = this.activated[key];
      if (value === true) {
        this.GdprObservable.active(key);
      }
    }
  }

  /**
   * @description init cookie settings if not exist
   * @return {void}
   */
  initCookie(): void {
    const _c = this.cookie.getCookie();
    // if not cookie rgpd clear cookie
    if (_c === undefined) {
      this.clearCookies();
    }
    // create cookie object
    const value = this.createActivatedObject(_c);
    // save
    this.cookie.updateCookie(value);
  }

  /**
   * @description clear cookie
   * @return {void}
   */
  clearCookies(): void {
    this.cookie.removeAll(this.options.keepCookies);
  }

  /**
   * @description create object activated with old value if exist
   * @param {object | void} _c
   * @return {object}
   */
  createActivatedObject(_c: Object | void): Object {
    const {types} = this.options;
    // old value in cookie if exist
    const oldActivated = _c !== undefined ? _c : {};
    const newActivated = {};
    // if type is array
    if (types !== undefined && Array.isArray(types)) {
      types.forEach(name => {
        // old value or true by default
        newActivated[name] =
          oldActivated[name] !== undefined ? oldActivated[name] : false;
      });
    }
    this.activated = newActivated;
    return this.activated;
  }
}
