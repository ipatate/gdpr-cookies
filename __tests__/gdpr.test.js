import Gdpr from '../src/services/Gdpr';

test('test get existing option value', () => {
  const gdpr = new Gdpr({name: 'lol'});
  const optionName = gdpr.getOption('name');
  expect(optionName).toBe('lol');
});

test('test get not existing option value', () => {
  const gdpr = new Gdpr({name: 'lol'});
  const optionName = gdpr.getOption('foo');
  expect(optionName).toBeUndefined();
});

test('verify value default gdpr cookie', () => {
  const gdpr = new Gdpr();
  expect(gdpr.createActivatedObject()).toMatchSnapshot();
});

test('verify update value with old setting gdpr cookie', () => {
  const gdpr = new Gdpr({type: ['ads', 'stats']});
  const activated = gdpr.createActivatedObject({ads: true, foo: 'lol'});
  expect(activated.ads).toBeTruthy();
  expect(activated.stats).toBeFalsy();
});

test('recup global rgpd', () => {
  window._gdpr = [['foo', 'bar', () => {}]]; // eslint-disable-line
  const gdpr = new Gdpr({type: ['ads', 'stats']});
  expect(gdpr.getGlobalGdpr()).toMatchSnapshot();
});
