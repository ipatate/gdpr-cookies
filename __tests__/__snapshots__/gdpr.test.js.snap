// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`recup global rgpd 1`] = `
Array [
  Array [
    "foo",
    "bar",
    [Function],
  ],
]
`;

exports[`verify value default gdpr cookie 1`] = `
Object {
  "ads": false,
  "stats": false,
}
`;
